### ISSUE TITLE
Conflicting definition of cortex m4 macros on Gcc5

### ISSUE THREAD

**Issue by offa** [offa/stm32-eth#1]
Ci builds of _**Gcc5**_ fail because of a **conflicting definition** of `__I` and  `__O`.

-------------------------------------

Cortex M4 files (`core_m4.h`) define `__I` / `__O` as follows:

```cpp
#ifdef __cplusplus
  #define   __I     volatile             /*!< Defines 'read only' permissions */
#else
  #define   __I     volatile const       /*!< Defines 'read only' permissions */
#endif
#define     __O     volatile             /*!< Defines 'write only' permissions */
```
*Same for `__O`.*

Those conflict with identifiers used internally by Gcc 5 (only).

**Comment by offa** [comment]
Same with `__O`.

**Comment by offa** [comment]
Workaround implemented by 02d059390d9eb216a44a700402617ca27c464555:

Cmake will check the compiler whether it's affected or not. If so, both macros (`__I` and `__O`) are undef'ed – CMake will print an infomessage to indicate the workaround.

Crosscompiling (target) builds are not tested and should fail since this would be an (severe) integration issue – this shoulnd't happen though, since only identifiers used by x86 are affected anyway.

The header `Platform.h` includes necessary Stm32 files and takes care of this. Therefore files should **_NOT_** include Stm32 headers  directly.
